server/src/main/java/org/candlepin/async/tasks/HypervisorUpdateJob.java

```diff
@@ -0,0 +1,532 @@
+            while ((len = in.read(buffer)) > 0) {
+                baos.write(buffer, 0, len);
+            }
+            return new String(baos.toByteArray(), StandardCharsets.UTF_8);
+        }
+        catch (IOException e) {
+            throw new AssertionError(e);
+        }
+    }
+    /*
+     * Create a new hypervisor type consumer to represent the incoming hypervisorId
+     */
+
+    private Consumer createConsumerForHypervisorId(String incHypervisorId, String reporterId,
+        Owner owner, String principal, Consumer incoming) {
+        Consumer consumer = new Consumer();
+        if (incoming.getName() != null) {
+            consumer.setName(incoming.getName());
+        }
+        else {
+            consumer.setName(sanitizeHypervisorId(incHypervisorId));
+        }
+        consumer.setType(hypervisorType);
+        consumer.setFact("uname.machine", "x86_64");
+        consumer.setGuestIds(new ArrayList<>());
+        consumer.setLastCheckin(new Date());
+        consumer.setOwner(owner);
+        consumer.setAutoheal(true);
+        consumer.setCanActivate(subAdapter.canActivateSubscription(consumer));
+        if (owner.getDefaultServiceLevel() != null) {
+            consumer.setServiceLevel(owner.getDefaultServiceLevel());
+        }
+        else {
+            consumer.setServiceLevel("");
+        }
+        if (principal != null) {
+            consumer.setUsername(principal);
+        }
+        consumer.setEntitlementCount(0L);
+        // TODO: Refactor this to not call resource methods directly
+        consumerResource.sanitizeConsumerFacts(consumer);
+
+
+        // Create HypervisorId
+        HypervisorId hypervisorId = new HypervisorId(consumer, owner, incHypervisorId);
+        hypervisorId.setReporterId(reporterId);
+        consumer.setHypervisorId(hypervisorId);
+
+        // TODO: Refactor this to not call resource methods directly
+        consumerResource.checkForFactsUpdate(consumer, incoming);
+
+        return consumer;
+    }
+    /*
+     * Make sure the HypervisorId is a valid consumer name.
+     */
+
+    private String sanitizeHypervisorId(String incHypervisorId) {
+        // Same validation as consumerResource.checkConsumerName
+        if (incHypervisorId.indexOf('#') == 0) {
+            log.debug("Hypervisor id cannot begin with # character");
+            incHypervisorId = incHypervisorId.substring(1);
+        }
+
+        int max = Consumer.MAX_LENGTH_OF_CONSUMER_NAME;
+        if (incHypervisorId.length() > max) {
+            log.debug("Hypervisor id too long, truncating");
+            incHypervisorId = incHypervisorId.substring(0, max);
+        }
+        return incHypervisorId;
+    }
+
+    /**
+     * Job configuration object for the hypervisor update job
+     */
+    public static class HypervisorUpdateJobConfig extends JobConfig {
+
+        public HypervisorUpdateJobConfig() {
+            this.setJobKey(JOB_KEY)
+                .setJobName(JOB_NAME);
```

Comment:
This job is currently configured to have unique-by-entity & other constraints. You can see that by checking the old jobs that either 1) extend `UniqueByEntityJob`, or 2) explicitly define the methods `isSchedulable` & `scheduleJob`. 

The old HypervisorUpdateJob is doing #2, with the target being the owner, but it is also using a custom query during the `isSchedulable` method that not only checks for non-terminal jobs of the same type against the same owner, that ALSO checks & schedules these jobs in the order they were received. So it will only allow the job to run if it is the oldest of its type against this owner (see change f15807dde0e4ca0f08fcee7b94e7b4b29327e478 for details).

Currently, the new way we do constraints is to add the constraint here, in similar way to the ImportJob/ExportJob: `.addConstraint(JobConstraints.uniqueByArgument(OWNER_KEY));`. This needs to be done now at a minimum, but it not take care the requirement of preserving the order. This is another topic in the list of things we need to discuss when @Ceiu comes back I think.